### PR TITLE
Fix problems discovered by numpy with type hints.

### PR DESCRIPTION
**Pull request by jesnie**
Basically various fixes to allow us to run with both older and newer versions of `{Python, TensorFlow, numpy, mypy}`.

### CODE REVIEW

File: gpflow/base.py [hunk 1]
```diff
@@ -26,12 +26,22 @@
     from IPython.lib import pretty
 
 DType = Union[np.dtype, tf.DType]
-VariableData = Union[List[Any], Tuple[Any], np.ndarray, int, float]  # deprecated
+
+if TYPE_CHECKING:
+    # It would be nice to use something more interesting than `Any` here, but it looks like the
+    # infrastructure in the rest of the ecosystem isn't really set up for this yet. Maybe when we
+    # get Python 3.11?
+    AnyNdArray = Union[np.ndarray[Any, Any]]
```
Comment (uri-granta) [inline gpflow/base.py:34]: I ended up using just writing `"np.ndarray[Any, Any]"` when I needed it (with quotes), which works both at type checking and runtime, but this is probably nicer. Fortunately we have very few ndarrays in trieste, as it's nearly all tensors.

File: gpflow/expectations/expectations.py [hunk 2]
```diff
@@ -121,7 +121,8 @@ def _init_expectation(
     obj1, feat1 = obj1 if isinstance(obj1, tuple) else (obj1, None)
     obj2, feat2 = obj2 if isinstance(obj2, tuple) else (obj2, None)
     return (
-        cast(ProbabilityDistribution, p),
+        # type-ignore instead of cast, because it dependes on versions whether a cast is necessary.
+        p,  # type: ignore
```
Comment (uri-granta) [inline gpflow/expectations/expectations.py:125]: Do you not get an unnecessary type ignore error then in one of the versions? Update: just spotted the config change.

File: gpflow/base.py (hunk 1 above)
Comment (uri-granta) [reply gpflow/base.py:34]: Also, to be consistent with `numpy.typing.NDArray`, you may want to name this `AnyNDArray`

File: gpflow/models/util.py [hunk 3]
```diff
@@ -12,20 +12,26 @@
...
 from .training_mixins import Data, ExternalDataTrainingLossMixin
 
+_I = TypeVar("_I", bound=InducingVariables)
+_InducingILike = Union[_I, tf.Tensor, AnyNdArray]
 
-def inducingpoint_wrapper(
-    inducing_variable: Union[InducingVariables, tf.Tensor, np.ndarray]
-) -> InducingVariables:
+
+InducingVariablesLike = _InducingILike[InducingVariables]
```
Comment (uri-granta) [inline gpflow/models/util.py:30]: Do you use `_InducingILike` anywhere else? If not, isn't it less confusing to just define:

```python
InducingVariablesLike = Union[InducingVariables, tf.Tensor, AnyNdArray]
InducingPointsLike = Union[InducingPoints, tf.Tensor, AnyNdArray]
```

File: tests/gpflow/models/test_variational.py [hunk 4]
```diff
@@ -30,38 +31,44 @@
...
+    meanA: AnyNdArray, meanB: AnyNdArray, varA: AnyNdArray, varB: AnyNdArray
+) -> AnyNdArray:
     # KL[ qA | qB ] = E_{qA} \log [qA / qB] where qA and qB are univariate normal distributions.
     return 0.5 * (
-        np.log(varB) - np.log(varA) - 1.0 + varA / varB + (meanB - meanA) * (meanB - meanA) / varB
+        np.log(varB)
+        - np.log(varA)
+        - 1.0
+        + varA / varB
+        + cast(AnyNdArray, meanB - meanA) * cast(AnyNdArray, meanB - meanA) / varB
```
Comment (uri-granta) [inline tests/gpflow/models/test_variational.py:57]: (why is this necessary? what does mypy think otherwise?)

**Review by uri-granta — APPROVED** [review]
Not too painful!

File: gpflow/models/util.py (hunk 3 above)
Comment (jesnie) [reply gpflow/models/util.py]: > Do you use `_InducingILike` anywhere else? If not, isn't it less confusing to just define:
> 
> ```python
> InducingVariablesLike = Union[InducingVariables, tf.Tensor, AnyNdArray]
> InducingPointsLike = Union[InducingPoints, tf.Tensor, AnyNdArray]
> ```

Sure. I used to use it, but then stuff didn't work, and I guess I don't need it anymore.

File: gpflow/expectations/expectations.py (hunk 2 above)
Comment (jesnie) [reply gpflow/expectations/expectations.py:125]: > Do you not get an unnecessary type ignore error then in one of the versions? Update: just spotted the config change.

:+1:  Yes, there are a couple of places where I've given up on writing code that is type-correct with all versions of numpy.
